Create new CLI

dtext opened the pull request:
This is a completely different, more standard command line interface. Using `docopt` instead of `argparse` for the most readable solution. Different features are seperated into different commands to improve readability, extensibility and maintainability. 

Closes #18.

dtext committed:
Add first implementation of proper CLI using docopt

dtext committed:
Implement logging

dtext committed:
Port screenshot functionality from instantshare.py over to screen command

dtext committed:
Port tray functionality over from instantshare.py to gui command

dtext committed:
Add missing log message in screen.py

dtext committed:
Fix working directory issue.

Can't wait for the resolution of #20

dtext committed:
Remove irrelevant TODO

dtext committed:
Integrate new CLI

dtext commented on src/cli/main.py:
```diff
@@ -0,0 +1,79 @@
+def execute_command(cmd, argv=None, *args, **kwargs):
+    if not argv:
+        argv = list(args) + ["--{}={}".format(k, v) for k, v in kwargs]
+    _get_module(cmd).main(argv)
```
I could definitely also change this function to just take the required `argv` for a specific command.
That would make it simpler and allow us to do things like `execute_command(shlex.split("screen --whole"))`.
Right now, this function does a bit of everything, which isn't necessarily ideal. Let me know what you think.

denniseffing replied:
You should consider two functions: 
1. `_execute_command` for internal usage (in cli/main.py) with `argv` as argument (`argv[0]` contains the command)
2. `execute_command` for easier external usage (in cli/<command>.py) with a string as argument, calling _execute_command(shlex.split(string))`.

Then use `_execute_command(argv)` in line 70.

dtext replied:
I agree with you. Will try to come up with a better design.

denniseffing replied:
The current implementation also contains another bug which causes a crash when using the menu options of the tray icon: You didn't include `cmd` in `argv` when the `argv` function parameter is `None`.

Fix:

``` python
if not argv:
    argv = [cmd] + list(args) + ["--{}={}".format(k, v) for k, v in kwargs]
```

denniseffing commented on src/cli/main.py:
```diff
@@ -0,0 +1,79 @@
+"""
+Usage: instantshare.py <command> [<args>...]
+       instantshare.py [-h | --help | -v | --version]
+       instantshare.py help <command>
```
Change the order to 

```
Usage: instantshare.py [-h | --help | -v | --version]
           instantshare.py <command> [<args>...]
           instantshare.py help <command>
```

dtext replied:
> Change the order

I don't see the advantage of that. The topmost command is the most commonly used one, so that should be at the top. After all, you don't type `instantshare --help` just to find out what you need to type to get help.

> change the name of the program to "instantshare"

Would have already done that, just wasn't sure whether or not python on windows would handle that, as Windows tends to be picky about filenames. So thanks for clarifying!

denniseffing commented on src/cli/main.py:
```diff
@@ -0,0 +1,79 @@
+        print("instantshare version 0.1")
+    elif args["<command>"] == "help":
+        cmd = args["<args>"][0]
+        print(_get_module(cmd).__doc__)
```
Causes `IndexError: list index out of range` when calling instantshare with "python instantshare.py help"

dtext replied:
@denniseffing: In fact, the `IndexError` is already raised in line 65, which is why I didn't get that immediately. Should be fixed in 7ec5c53 though.

denniseffing commented on src/cli/main.py:
```diff
@@ -0,0 +1,79 @@
+"""
+Usage: instantshare.py <command> [<args>...]
+       instantshare.py [-h | --help | -v | --version]
+       instantshare.py help <command>
```
As you already mentioned in #18, we want to create an executable file for Windows in the long run, so you should change the name of the program to "instantshare" instead of "instantshare.py" and rename the file "instantshare.py" to "instantshare". This prevents inconsistency with the usage documentation later on when using different platforms.

The shebang `#!/usr/bin/env python3` allows executing the script with the correct interpreter on Linux by just executing it and for now it is also possible to execute the script on windows by using "python instantshare" (even without the .py file extension).

dtext committed:
Small fixes and refactoring

dtext committed:
Rename instantshare.py to instantshare

dtext committed:
Print usage pattern to stderr and exit with error code when user did not specify a command after 'help'

dtext committed:
Fix mistake in error condition

dtext commented:
I think we should be good now, right?

Edit: forgot one thing, everything should be fine now.

dtext committed:
Also parse arguments in gui command

Every subcommand needs to parse its argv using docopt so 'instantshare
help <command>' works correctly. If we don't do it this way, the main
method will be executed regularly.

denniseffing commented:
:+1:
